### PR TITLE
Add eventlog, package for semi-structured event logging

### PR DESCRIPTION
**Pull request by jcharum**
Add the `eventlog` package which provides semi-structured event logging for convenient analytics.  Example usage:
```
sess := session.NewSession()
cw := cloudwatchlogs.New(sess)
e := NewCloudWatchEventer(cw, "myLogGroup", "myLogStream")
e.Event("rpcRetry", "org", "GRAIL", "retry", 0, "maxRetry", 10)
e.Event("machineStopped", "addr", "192.168.1.1", "duration", 3600.0, "startTime": 1584140534)
```

These events will then be available in CloudWatch Logs (and therefore easily made available in Athena) for analysis.

### CODE REVIEW

**Comment by jcharum** [comment]
See potential usage in [this bigmachine branch](https://github.com/grailbio/bigmachine/compare/master...jcharum:eventlog?expand=1).

File: eventlog/cloudwatch/cloudwatch.go [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+
+// syncInterval is the maximum duration that we will wait before sending any
+// buffered messages to CloudWatch.
+const syncInterval = 1 * time.Second
+
+// eventBufferSize is size of the channel buffer used to process events. When
+// this buffer is full, new events are dropped.
+const eventBufferSize = 32768
+
+func init() {
```
Comment (jacklionheart) [inline eventlog/cloudwatch/cloudwatch.go:45]: Is there any use in some sort of `close` or `onExit` style hook? I imagine we want to throw events whenever bad things happen. What's the best way to ensure we lose as little as possible?

File: eventlog/eventlog.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+			return Nop{}, nil
+		}
+	})
+	// We do the most conservative thing by default, making event logging a
+	// no-op.
+	config.Default("eventer", "eventer/nop")
+}
+
+// Eventer is called to log events.
+type Eventer interface {
```
Comment (jacklionheart) [inline eventlog/eventlog.go:37]: I thought I might see some sort of Eventer mock that implements this interface and can be used during tests that just stores the logged events locally for inspection during testing.

File: eventlog/cloudwatch/cloudwatch.go [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+		client:    client,
+		group:     group,
+		stream:    stream,
+		eventc:    make(chan event, 1024),
+		syncc:     make(chan struct{}),
+		syncDonec: make(chan struct{}),
+		donec:     make(chan struct{}),
+	}
+	var ctx context.Context
+	ctx, eventer.cancel = context.WithCancel(context.Background())
```
Comment (jacklionheart) [inline eventlog/cloudwatch/cloudwatch.go:120]: Would adding an explicit timeout help?

File: eventlog/cloudwatch/cloudwatch.go (hunk 1 above)
Comment (jacklionheart) [reply eventlog/cloudwatch/cloudwatch.go:45]: It does look like you're buffering, so curious about that.

File: eventlog/eventlog.go (hunk 2 above)
Comment (jacklionheart) [inline eventlog/eventlog.go:37]: Do we expect this functionally to be used only by bigslice internal code or by users of bigslice? Do we expect to mainly use this base API, or will we want any additional layers on top. For example, we could have different functions that add some common useful k:v pairs.

File: eventlog/cloudwatch/cloudwatch.go [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+// string of CloudWatch Log event.
+const maxSingleMessageSize = maxBatchSize - 26
+
+// syncInterval is the maximum duration that we will wait before sending any
+// buffered messages to CloudWatch.
+const syncInterval = 1 * time.Second
+
+// eventBufferSize is size of the channel buffer used to process events. When
+// this buffer is full, new events are dropped.
+const eventBufferSize = 32768
```
Comment (jacklionheart) [inline eventlog/cloudwatch/cloudwatch.go:43]: Is this used?

File: eventlog/cloudwatch/cloudwatch.go [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+		}
+		if len(s) > maxSingleMessageSize {
+			log.Error.Printf("CloudWatchEventer: dropping log event: message too large")
+			return
+		}
+		newBatchSize := batchSize + len(s) + 26
+		if newBatchSize > maxBatchSize {
+			sync(true)
+		}
+		inputLogEvents = append(inputLogEvents, &cloudwatchlogs.InputLogEvent{
```
Comment (jacklionheart) [inline eventlog/cloudwatch/cloudwatch.go:236]: It looks like this is all one thread and doesn't require any locking, right?

File: eventlog/cloudwatch/cloudwatch.go [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+	go eventer.loop(ctx)
+	return eventer
+}
+
+// Event implements Eventer.
+func (c *CloudWatchEventer) Event(typ string, fieldPairs ...interface{}) {
+	select {
+	case c.eventc <- event{timestamp: time.Now(), typ: typ, fieldPairs: fieldPairs}:
+		atomic.StoreInt32(&c.loggedFullBuffer, 0)
+	default:
```
Comment (jacklionheart) [inline eventlog/cloudwatch/cloudwatch.go:130]: What's going on here?

**Review by jacklionheart — APPROVED** [review]
Looks great. Please do consider my feedback, but I think it can be addressed in future commits, so accepting.

File: eventlog/cloudwatch/cloudwatch.go (hunk 1 above)
Comment (jcharum) [reply eventlog/cloudwatch/cloudwatch.go:45]: The semantics are similar to other streams. You have to `Close` it to make sure data is flushed/synced. In general, this is all best-effort, and I think the short buffer interval (1 second) should general make it good enough. If clients do want to do more, it's up to them to set up to call `Close` appropriately, e.g. at exit, in a signal handler, etc. I'm inclined to not worry about it for now.

File: eventlog/eventlog.go (hunk 2 above)
Comment (jcharum) [reply eventlog/eventlog.go:37]: Yes; right now I think of this more like (best-effort) logging, just formatted for easy parsing and with data we're specifically looking to collect. I don't expect people to test the behavior just like they don't test that their programs write logs properly, even if there are scripts out there that do some ad hoc parsing of logs. If it evolves into something on which rely more critically, we'll definitely have to revisit.

File: eventlog/cloudwatch/cloudwatch.go (hunk 3 above)
Comment (jcharum) [reply eventlog/cloudwatch/cloudwatch.go:120]: This context exists to stop the loop that processes events and writes them out to CloudWatch Logs. It is only canceled on `Close`.

File: eventlog/eventlog.go (hunk 2 above)
Comment (jcharum) [reply eventlog/eventlog.go:37]: My plan is for this to be just used by bigmachine and bigslice for now to give us time to work out possible operational issues, etc. If we find that the mechanism is very useful, I'd invite others to start using it, though with their own `Eventer` instances. At that point, we might want to expand the API, but I'd wait until then.

As always, my plans are up for discussion, so let me know if you have different ideas.

File: eventlog/cloudwatch/cloudwatch.go (hunk 4 above)
Comment (jcharum) [reply eventlog/cloudwatch/cloudwatch.go:43]: I meant to use it, but I accidentally did not. A future commit will remedy this.

File: eventlog/cloudwatch/cloudwatch.go (hunk 5 above)
Comment (jcharum) [reply eventlog/cloudwatch/cloudwatch.go:236]: That's right; there is only one goroutine accessing `inputLogEvents`, the one running `loop`.

File: eventlog/cloudwatch/cloudwatch.go (hunk 6 above)
Comment (jcharum) [reply eventlog/cloudwatch/cloudwatch.go:130]: I attempted to explain this in the comment for `loggedFullBuffer`:
```
// loggedFullBuffer prevents consecutive printing of "buffer full" log
// messages inside Event. We get a full buffer when we are overloaded with
// many messages. Logging each dropped message is very noisy, so we suppress
// consecutive logging.
```

Does that make sense?

**Pull request merged**